### PR TITLE
feat: add mergeTableConfig() for composing base + override configs

### LINKED ISSUE #337: feat: add mergeTableConfig() utility for composing base + override configs
## What

Export a `mergeTableConfig(base, overrides)` utility that deep-merges `strategies` sub-keys while shallow-merging everything else — the pattern every project needs when building a shared base config with per-table overrides.

## Why

Without it, users write manual spread chains where order is load-bearing and easy to get wrong:

\`\`\`js
return useTable(tableLocator, {
  ...baseConfig,
  ...overrides,
  strategies: {
    ...overrides.strategies,
    loading: {
      ...baseConfig.strategies.loading,  // base defaults
      ...(overrides.strategies?.loading || {}),  // caller wins
    },
    // every new default strategy needs another manual merge layer
  },
});
\`\`\`

Adding a second default strategy (e.g. a shared `pagination`) means another manual merge. Getting the spread order wrong silently clobbers one side.

## Proposed API

\`\`\`js
import { mergeTableConfig } from '@rickcedwhat/playwright-smart-table';

const grafanaBase = {
  rowSelector: '[role="row"][class*="row"]',
  maxPages: 100,
  strategies: {
    loading: {
      isCellLoading: async (cell) => ...,
      cellLoadingTimeout: 45_000,
      onCellLoadingTimeout: 'read-as-is',
    },
  },
};

function getGrafanaTable(page, { tableLocator, ...overrides }) {
  return useTable(tableLocator, mergeTableConfig(grafanaBase, overrides));
}
\`\`\`

## Merge semantics

- Top-level keys (`rowSelector`, `maxPages`, etc.): overrides win, same as `{ ...base, ...overrides }`
- `strategies`: each sub-key (`loading`, `pagination`, `sorting`, etc.) merged independently — override sub-key wins per property, base sub-key fills in the rest
- `columnOverrides`: merged key-by-key (same as strategies)
- Arrays (e.g. a future `plugins` key): overrides replace, not concat

## Notes

- Pure function, no side effects
- Should be TypeScript-friendly — return type is the merged config type
- Does not need to handle deep nesting beyond `strategies.*` — one level of deep merge is sufficient for all known use cases

### PR DESCRIPTION
**Pull request by rickcedwhat-ai**
## Summary

- Adds `mergeTableConfig(base, overrides)` utility to `src/utils/mergeTableConfig.ts`
- Exports it from `src/index.ts` as a named export
- Deep-merges `strategies` sub-objects (`loading`, `viewport`) key-by-key — neither side silently wipes the other
- Function-valued strategies (`getCellLocator`, `beforeCellRead`, etc.) are replaced by the override (can't be object-spread)
- `columnOverrides` merged entry-by-entry
- All other top-level keys: overrides win (shallow merge)
- 21 unit tests covering all merge paths, mutation safety, and edge cases
- `CHANGELOG.md` updated under `[Unreleased]`

Closes #337.

## Test plan

- [x] `npx vitest run tests/unit/mergeTableConfig.test.ts` — 21 tests pass
- [x] `npx tsc --noEmit` — no type errors
- [x] `pnpm run build` — passes (run by pre-commit hook)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a new configuration merge utility for combining a shared base setup with per-table overrides.
  * The merge behavior now preserves existing settings while letting overrides take precedence, including deeper handling for strategy and column-specific settings.

* **Documentation**
  * Updated the changelog to note the new merge utility.

* **Tests**
  * Added coverage for merge behavior, override precedence, empty inputs, and ensuring original inputs are not modified.

### CODE REVIEW

**Comment by rickcedwhat-ai** [comment]
## 🤖 Bot HQ


### 🔗 Issue Link
🔗 Closes #337


---




### 🔍 AI Review
- [ ] 🔍 Request AI review

> ✅ Review complete — no blocking issues.

📊 This PR: $0.0776 · Repo today: $0.12 / $1.00 · Month: $1.62 / $10.00

<details>
<summary><strong>Review History (2 rounds)</strong></summary>

| Round | Date | Commit | In Tokens | Out Tokens | Cost | Summary |
|---|---|---|---|---|---|---|
| 1 | 2026-06-25 | `7d32bb5` | 37,852 | 108 | $0.0384 | No issues found |
| 2 | 2026-06-25 | `611ea7b` | 38,718 | 105 | $0.0392 | No issues found |


</details>


---
<sub>This comment is managed by the bot — do not edit directly.</sub>

**Review by rickcedwhat — APPROVED** [review]

**Comment by rickcedwhat-ai** [comment]
## AI Review

PR #341 adds `mergeTableConfig()` utility for composing base and override table configs with smart deep-merging of strategies and columnOverrides. All 21 tests pass, TypeScript is strict, and the implementation is sound with no logic errors or security issues found.
> ✅ No actionable issues found.

File: src/utils/mergeTableConfig.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * Deep-merges two `TableConfig` objects.
+ *
+ * Merge semantics:
+ * - Top-level keys (`rowSelector`, `maxPages`, etc.): overrides win — same as `{ ...base, ...overrides }`.
+ * - `strategies`: each sub-key is merged independently. Strategy sub-objects (`loading`, `viewport`)
+ *   are merged key-by-key so that neither side silently wipes the other's settings.
+ *   Function-valued strategies (`pagination`, `header`, `dedupe`, `getCellLocator`, etc.) are
+ *   replaced in full by the override value (a function cannot be meaningfully object-spread).
```
[the remark this answers is not in the document]
Comment (rickcedwhat-ai) [reply src/utils/mergeTableConfig.ts]: Fixed — docstring now correctly states that object-valued strategies (`loading`, `viewport`, `pagination`) are merged key-by-key, and function-valued strategies (`header`, `dedupe`, `getCellLocator`, etc.) are replaced in full.

File: src/utils/mergeTableConfig.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                baseVal !== undefined &&
+                overrideVal !== undefined &&
+                typeof baseVal === 'object' &&
+                typeof overrideVal === 'object' &&
+                !Array.isArray(baseVal) &&
+                !Array.isArray(overrideVal) &&
+                typeof baseVal !== 'function' &&
+                typeof overrideVal !== 'function'
+            ) {
+                (result.strategies as any)[key] = { ...baseVal, ...overrideVal };
```
[the remark this answers is not in the document]
Comment (rickcedwhat-ai) [reply src/utils/mergeTableConfig.ts]: Fixed — cast is now `(result.strategies as NonNullable<TableConfig<T>['strategies']>)[key] = { ...baseVal, ...overrideVal } as never`. The outer cast is typed; only the spread value uses `as never`, which is the standard idiom for a union assignment TypeScript can't express but that the runtime object-check above guarantees is safe.

File: tests/unit/mergeTableConfig.test.ts [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            const overrides: TableConfig = {
+                strategies: { pagination: overridePagination as any },
+            };
+
+            const result = mergeTableConfig(base, overrides);
+            // pagination is a plain object, so it should be merged like other plain objects
+            // Actually pagination is an object with function values — it is itself an object
+            // and should be deep-merged. Let's verify override wins the goNext key:
+            expect(result.strategies?.pagination).toBeDefined();
+        });
```
[the remark this answers is not in the document]
Comment (rickcedwhat-ai) [reply tests/unit/mergeTableConfig.test.ts]: Fixed — assertion now verifies `result.strategies?.pagination?.goNext` is the override's function and `result.strategies?.pagination?.goNextBulk` is preserved from base, replacing the weak `toBeDefined()` check.

**Comment by rickcedwhat-ai** [comment]
## AI Review — Round 2

PR #341 adds mergeTableConfig() utility with deep-merge logic for composing base and override configs. The implementation is sound with comprehensive unit tests (21 passing) and proper export setup. No type correctness issues, security vulnerabilities, or logic errors found.
> ✅ No actionable issues found.
